sim/midas/src/main/scala/midas/core/FPGATop.scala

```diff
@@ -225,32 +276,54 @@ class FPGATop(implicit p: Parameters) extends LazyModule with HasWidgets {
   val toCPUStreamParams = bridgesWithToHostCPUStreams.map { _.streamSourceParams }
   val fromCPUStreamParams = bridgesWithFromHostCPUStreams.map { _.streamSinkParams }
 
-  val pcisAXI4BundleParams = AXI4BundleParameters(
-    addrBits = p(DMANastiKey).addrBits,
-    dataBits = p(DMANastiKey).dataBits,
-    idBits   = p(DMANastiKey).idBits)  // Dubious...
+  val streamingEngine = addWidget(p(StreamEngineInstantiatorKey)(
+    StreamEngineParameters(toCPUStreamParams.toSeq, fromCPUStreamParams.toSeq), p)
+  )
+
+  require(streamingEngine.fpgaManagedAXI4NodeOpt.isEmpty || p(FPGAManagedAXI4Key).nonEmpty,
+    "Selected StreamEngine uses the FPGA-managed AXI4 interface but it is not available on this platform."
+  )
+  require(streamingEngine.cpuManagedAXI4NodeOpt.isEmpty || p(CPUManagedAXI4Key).nonEmpty,
+    "Selected StreamEngine uses the CPU-managed AXI4 interface, but it is not available on this platform."
+  )
 
-  val pcisNode = AXI4MasterNode(
-    Seq(AXI4MasterPortParameters(
+  val cpuManagedAXI4NodeTuple =  p(CPUManagedAXI4Key).map { params =>
+    val node = AXI4MasterNode(Seq(AXI4MasterPortParameters(
       masters = Seq(AXI4MasterParameters(
-          name       = "cpu-mastered-axi4",
-          id         = IdRange(0, 1 << p(DMANastiKey).idBits),
+          name       = "cpu-managed-axi4",
+          id         = IdRange(0, 1 << params.idBits),
           aligned    = false,
-          maxFlight  = None, // None = infinite, else is a per-ID cap
+          maxFlight  = params.maxFlight, // None = infinite, else is a per-ID cap
         ))
       )
-    )
-  )
+    ))
+    streamingEngine.cpuManagedAXI4NodeOpt.foreach {
+      _ := AXI4Buffer() := node
+    }
+    (node, params)
+  }
 
-  val streamingEngine = addWidget(p(StreamEngineInstantiatorKey)(
-    StreamEngineParameters(toCPUStreamParams.toSeq, fromCPUStreamParams.toSeq), p)
-  )
+  val fpga_managed_axi4NodeTuple: Option[(AXI4SlaveNode, FPGAManagedAXI4Params)] =  p(FPGAManagedAXI4Key).map { params =>
```

Comment:
Why is the type specified for `fpga_managed_axi4NodeTuple` but not for the `cpu...` equivalent?

Reply:
I'll make this consistent.